### PR TITLE
[expo-updates] Update spec for version 1: message part

### PR DESCRIPTION
**Pull request by wschurman**
# Why

One new feature coming soon to the expo-updates client library is support for rollbacks. A rollback is a directive from the server telling the client to load the embedded update (the one included in the built app) instead of any more recently-downloaded remote updates.

To accomplish this, a new method must be created for communicating this directive. Since the only protocol used for communicating between server and client is the update request, the new directive should be communicated over the update request.

This PR adds the specification for a new "part" to multipart update requests, "message". It is underspecified on purpose in this spec since compliant servers and clients can use this for any application needs. The "message" part will be consumed by compliant client implementations alongside other update parts. A compliant client implementation should consume the message and may choose to do things like modify behavior of an update, change global update state, or ignore the message entirely. This PR just adds the protocol definition and version change necessary to add it as an optional manifest part.

# How

Add new specification. Note that this was mv'd from the previous specification to show a clean diff, but I will archive a copy of the version 0 of the specification before landing this.

# How reviewers can help

1. Should this version of the protocol just eliminate non-multipart entirely?
    - It muddies the new message addition since it is only supported in multipart responses (versus manifests which were compatible with both formats of responses).
    - It's unlikely that we'll be able to remove non-multipart from the EAS servers but that doesn't mean that this new version needs to include it. Instead it means that the EAS server will just support multiple versions of the protocol.
2. Should the message types be specified in this specification? My gut said that the level of abstraction defined in this protocol didn't dictate how a client implementation needed to function, but another part of me thinks that maybe rollbacks should be a part of this protocol?
3. How should we think about adding new message types in the future? My gut says that we should just specify that unknown message types should be ignored by clients, but then how will the server know whether to send them? Does adding a new message type require bumping the spec version? Should we add another header like `expo-expect-message-types`?
4. Should we support multiple messages?

# Test Plan

Read the specification.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (douglowder) [inline docs/pages/technical-specs/expo-updates-1.mdx:1]: If you are going to do this, you need to also modify `docs/constants/navigation.js` to navigate to `expo-updates-1` instead of `expo-updates-0`

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (douglowder) [inline docs/pages/technical-specs/expo-updates-1.mdx:41]:
```suggestion
    - For a _message_ part, the client library will consume the message depending on the message type and update its local state accordingly.
```

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (douglowder) [inline docs/pages/technical-specs/expo-updates-1.mdx:58]: Not sure what is meant by `based on the supported response structures though SHOULD send` -- can you clarify this?

**Review by douglowder — CHANGES_REQUESTED** [review]
See inline comments

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (Simek) [reply docs/pages/technical-specs/expo-updates-1.mdx:1]: We also will need to add redirect in:
* https://github.com/expo/expo/blob/main/docs/deploy.sh
* https://github.com/expo/expo/blob/main/docs/common/error-utilities.ts

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (douglowder) [reply docs/pages/technical-specs/expo-updates-1.mdx]: Let's discuss this on Monday to make sure I understand it :)

Comment (wschurman) [reply docs/pages/technical-specs/expo-updates-1.mdx]: Essentially a compliant client can choose to only support non-multipart for simplicity (for example, if a consumer of the protocol only needs the manifest and doesn't need any of the other stuff). This sentence is saying that it has the option, but it is suggested to support multipart responses since there is more that can be done with those responses.

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (douglowder) [inline docs/pages/technical-specs/expo-updates-1.mdx:58]:
```suggestion
A conformant client library MUST send at least one of `accept: application/expo+json`, `accept: application/json`, or `accept: multipart/mixed` based on the [supported response structures](#update-response), though it SHOULD send `accept: application/expo+json, application/json, multipart/mixed`. A conformant client library MAY express preference using "q" parameters as specified in [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.1), which default to `1`.
```

**Comment by douglowder** [comment]
> Should this version of the protocol just eliminate non-multipart entirely? It muddies the new message addition since it is only supported in multipart responses (versus manifests which were compatible with both formats of responses). It's unlikely that we'll be able to remove non-multipart from the EAS servers but that doesn't mean that this new version needs to include it. Instead it means that the EAS server will just support multiple versions of the protocol.

I would support eliminating non-multipart.

> Should the message types be specified in this specification? My gut said that the level of abstraction defined in this protocol didn't dictate how a client implementation needed to function, but another part of me thinks that maybe rollbacks should be a part of this protocol?

> How should we think about adding new message types in the future? My gut says that we should just specify that unknown message types should be ignored by clients, but then how will the server know whether to send them? Does adding a new message type require bumping the spec version? Should we add another header like expo-expect-message-types?

Since this is intentionally a very general protocol, we should not specify any message types as required, and should not require that rollbacks be part of the protocol. I think the `expo-expect-message-types` header is the right way to deal with this. Specifically for EAS, we can require that `rollback` message type be included in that header for EAS update clients.

> Should we support multiple messages?

It seems to me that this is supported by implication, since a multipart response could contain multiple message parts. Is this what you meant by supporting multiple messages?

**Review by douglowder — DISMISSED** [review]

**Comment by wschurman** [comment]
> > Should we support multiple messages?
> 
> It seems to me that this is supported by implication, since a multipart response could contain multiple message parts. Is this what you meant by supporting multiple messages?

It's more a question if we should support multiple "message" multipart parts (or have the "message" part be an array of messages). So a server could send back two directives in a single message, for example rollbackToEmbedded and updateServerDefinedHeaders and both would be executed.

I think we've settled on just allowing one for this version of the protocol.

**Comment by ide** [comment]
Here's my "keep it simple" proposal: there can be one part with one directive. In the future, we can either add support for multiple parts or a "multidirective" that takes an array of directives as the payload.

Practically speaking right now I think we have four ideas for directives: roll back to an embedded update (potentially with the ID of that embedded update specified as a parameter), a no-op (though perhaps we should make the spec allow sending back no body and a 204 response code to be a no-op instead of creating a no-op directive), setting server-defined headers, and toggling diagnostics maybe. What would be the most ergonomic way to specify multiple of those, if we think they should be directives (I could see toggling diagnostics to be a header though)?

**Comment by wschurman** [comment]
I'll rebase this and move the files and fix the links once it has been approved (keeping it in this state until then to ease review since github is terrible at generating diffs otherwise).

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (ide) [inline docs/pages/technical-specs/expo-updates-1.mdx:220]: Since all three multipart parts are optional, what if instead of a `noUpdateAvailable` directive, we had the spec treat an empty multipart response as a no-op? That would also save on response bytes. The client should also accept either a 204 or 200 in that case.

Comment (ide) [inline docs/pages/technical-specs/expo-updates-1.mdx:220]: Is the directive type meant to be a verb phrase (roll back) or a noun phrase (a rollback)?

```suggestion
A conformant client library and server SHOULD specify and implement directive types specific to the needs of the application. For example, Expo Application Services makes use of two directive types: `noUpdateAvailable` (directs the expo-updates library to proceed as normal) and `rollBackToEmbedded` (directs the expo-updates library to use the update embedded in the host application instead of any other downloaded updates).
```

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (ide) [inline docs/pages/technical-specs/expo-updates-1.mdx:213]: As I understand it, these extra fields are allowed for forwards-compatibility. The spec should say to allow these fields and to ignore any that aren't documented by the spec so that a client doesn't start relying on unspecced fields.

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (ide) [inline docs/pages/technical-specs/expo-updates-1.mdx:212]: Usually I like context-insensitive names in codebases but in a protocol the shorter names fit better IMO, what do you think?
```suggestion
  type: string;
  data?: { [key: string]: any };
```

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (ide) [inline docs/pages/technical-specs/expo-updates-1.mdx:34]: If a manifest has been saved but the entire update hasn't been fully downloaded, we don't want to load that manifest.

Separately (unrelated to this PR), "cache" is the wrong word here in that a cache implies the data can be arbitrarily erased and repopulated, but in our case we want to make sure that updates are stable.
```suggestion
An app running a conformant Expo Updates client library MUST load the most recent update saved in the client library's update database possibly after filtering by the contents of the manifest's [_metadata_](#manifest-body).
```

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (ide) [inline docs/pages/technical-specs/expo-updates-1.mdx:31]: The spec goes on to talk about multipart responses but not directive responses -- probably need to fix something here.

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (ide) [inline docs/pages/technical-specs/expo-updates-1.mdx:28]: As explained in the feedback summary, I think directives should not be considered updates. An update is an object the client app launches in order to run, hence `update = (manifest, assets)`

**Review by ide — CHANGES_REQUESTED** [review]
The terminology needs to be straightened out, specifically around manifests and updates. Previously an update was defined as a unit consisting of a manifest and the assets needed to run the update.

Now, we have directive responses from the server. The current PR treats "manifest | directive" as "update" but I don't think that's quite right. A directive is not an update, it's an instruction to do something. It's not stored in the updates database.

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (wschurman) [reply docs/pages/technical-specs/expo-updates-1.mdx:213]: These extra fields are for things like code signing context (scopeKey, projectId) and anything else that a client-server implementation may need. Maybe we should specify an `extra` field as well with the same style as `manifest.extra` (for `scopeKey`, `eas.projectId`). 

Also, more generally this spec says: 

> A conforming implementation of this protocol MAY provide additional functionality, but MUST NOT where explicitly disallowed or would otherwise result in non-conformance.

So I think I'll specify `extra` in the spec and remove the `[key: string]: any;`.

Comment (ide) [reply docs/pages/technical-specs/expo-updates-1.mdx]: I do think it's a good idea to have the spec tell clients to allow and ignore unknown fields. It will help with forward compatibility if old clients will gracefully ignore fields instead of crashing if a new server sends back unknown fields.

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (ide) [inline docs/pages/technical-specs/expo-updates-1.mdx:216]: The word "payload" to me emphasizes that the data is being transported and is the body of a message. That's true here but these fields are more like parameters and arguments to the directive.

Some parts of the HTML spec (like Content Security Policies) use the terms "directive name" and "directive value". I think "value" is too generic and also seems a bit off when some of our update directives don't even need payloads/values/parameters.

Four words I considered are:
- parameters: I think this is best and adds a little precision by implying they parameterize the directive while also staying fairly generic and flexible
- data: OK but rather generic
- value: too generic and also implies a bit that it's necessary
- arguments: while maybe it's more technically correct than parameters, it's less intuitive than parameters and also is a reserved word in JS and can make coding a little less smooth

**Review by ide — CHANGES_REQUESTED** [review]

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (wschurman) [reply docs/pages/technical-specs/expo-updates-1.mdx:80]: I went with 406 and 406 here since returning a 200 in this case might mean we'd want to return it in all cases and that seems like it might be a bigger change.

**Review by ide — APPROVED** [review]

File: docs/pages/technical-specs/expo-updates-1.mdx
Comment (ide) [inline docs/pages/technical-specs/expo-updates-1.mdx:124]: Let's get this PR through but one thing we need to resolve in a follow-up is whether server-defined headers should be set when the server responds with an empty multipart response.If the client is expected to persist the server-defined headers, then no-op isn't the right word. 

We should take a look at expo-updates to see what it does today as well as our intent/motivation behind expo-server-defined-headers and then tidy up the spec. It's already kind of ambiguous about how to treat server-defined headers outside of the happy path where the server responds with a manifest.

File: docs/pages/bare/updating-your-app.mdx
Comment (douglowder) [inline docs/pages/bare/updating-your-app.mdx:108]:
```suggestion
This configuration is also generated by `npx expo prebuild` if you set the `requestHeaders` property in your **app.json**:
```

**Review by douglowder — APPROVED** [review]
One small nit otherwise looks good

**Pull request merged**
